url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/rtree/SecurityNode.java

```diff
@@ -0,0 +1,605 @@
+		}
+	}
+	
+	
+	public void writembrGfunctionArraylist(DataOutputStream dos, ArrayList<mbrGfunction> list) throws IOException{
+		dos.writeInt(list.size());
+		for(mbrGfunction mbrgf : list){
+			mbrgf.writeToFile(dos);
+		}
+	}
+	/**
+	 * load from integer data
+	 * 
+	 * @param data
+	 * @param rtree
+	 * @throws IOException 
+	 */
+	public SecurityNode(byte[] data, RTree rTree) throws IOException {
+		ByteArrayInputStream dais = new ByteArrayInputStream(data);
+		DataInputStream dis = new DataInputStream(dais);
+		childAttrHashValues = readStringArraylist(dis);
+		childGValues = readStringArraylist(dis);
+		//childmbrGfunctions = readmbrGfunctionArraylist(dis);
+		correspondingMBR = new long[2][2];
+		correspondingMBRids = new long[2][2];
+		for(int i = 0; i < 2; i++){
+			for(int j = 0; j < 2; j++){
+				correspondingMBR[i][j] = dis.readLong();
+				correspondingMBRids[i][j] = dis.readLong();
+			}
+		}
+		correspondNodeID = dis.readInt();
+		if (correspondNode == null) {
+			correspondNode = rTree.readNode(correspondNodeID);
+		}
+		entaValue = DataIO.readString(dis);
+		gValue = DataIO.readString(dis);
+		level = dis.readInt();
+		rsaentaValue = DataIO.readString(dis);
+		mbrDigest = DataIO.readString(dis);
+		childDigest = DataIO.readString(dis);
+	}
+
+	/**
+	 * save the SNode into Byte Array format:
+	 * <p>
+	 * childAttrHashValues, childIDs, correspondNodeID, entaValue, gValue, level
+	 * 
+	 * @return
+	 * @throws IOException 
+	 */
+	public byte[] saveTobytes() throws IOException {
+		ByteArrayOutputStream baos = new ByteArrayOutputStream();
+		DataOutputStream dos = new DataOutputStream(baos);
+		writeStringArraylist(dos, childAttrHashValues);
+		writeStringArraylist(dos, childGValues);
+		//writembrGfunctionArraylist(dos, childmbrGfunctions);
+		for(int i = 0; i < 2; i++){
+			for(int j = 0; j < 2; j++){
+				dos.writeLong(correspondingMBR[i][j]);
+				dos.writeLong(correspondingMBRids[i][j]);
+			}
+		}
+		dos.writeInt(correspondNodeID);
+		DataIO.writeString(dos, entaValue);
+		DataIO.writeString(dos, gValue);
+		dos.writeInt(level);
+		DataIO.writeString(dos, rsaentaValue);
+		DataIO.writeString(dos, mbrDigest);
+		DataIO.writeString(dos, childDigest);
+		dos.flush();
+		dos.close();
+		return baos.toByteArray();
+	}
+
+	/**
+	 * given a RTree node, this method outputs a correspond security node which
+	 * has a similar structure
+	 * 
+	 * @param rTree
+	 * @return
+	 * @throws Exception 
+	 * @throws IndexOutOfBoundsException 
+	 */
+	public static SecurityNode buildSecurityNode(SecurityTree srtree, RTree rTree) throws IndexOutOfBoundsException, Exception {
+		if (rTree == null) {
+			return null;
+		}
+		Node currentNode = rTree.readNode(rTree.m_rootID);
+		SecurityNode node = new SecurityNode();
+		DataOutputStream indexDos = new DataOutputStream(new FileOutputStream(srtree.filePath + ".sidx"));
+		DataOutputStream dataDos = new DataOutputStream(new FileOutputStream(srtree.filePath + ".sdat"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *dataDos*. The resource is referenced at lines 214, 215, 218. The resource is closed at line 219. There are other execution paths that do not contain closure statements, for example, when DataOutputStream.flush throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).